File: src/auth/oauth_server.rs
//! OAuth 本地回调 HTTP Server（需求 F-AUTH-02 / F-AUTH-06）。
//!
//! 对齐 `legacy/lib/auth/oauth_server.dart`。
//!
//! 绑定 127.0.0.1:port（不监听 0.0.0.0，满足安全要求）。
//! 监听 GET {callbackPath}，解析 code/state/error/sub_error。
//! 单次使用：拿到 code 后自动关闭。

use std::time::Duration;

use serde::Serialize;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpListener;
use tokio::sync::{oneshot, watch};
use tokio::task::JoinHandle;
use tokio::time::timeout;

use crate::constants;
use crate::error::{AppError, AppResult};

/// OAuth 回调结果
#[derive(Debug, Clone, Serialize, Default)]
pub struct OauthCallbackResult {
    pub code: Option<String>,
    pub state: Option<String>,
    /// 华为 error 码（如 '1101'）
    pub error: Option<String>,
    /// 华为 error_description（如 'invalid scope'）
    #[serde(rename = "errorDescription")]
    pub error_description: Option<String>,
    /// 华为 sub_error（如 '20042' 表示 scope 未授权）
    #[serde(rename = "subError")]
    pub sub_error: Option<String>,
}

impl OauthCallbackResult {
    /// 是否成功（有 code 且无 error）
    pub fn is_success(&self) -> bool {
        self.code.is_some() && self.error.is_none()
    }
}

/// 本地 OAuth 回调服务器。
///
/// 使用 tokio TcpListener 监听 127.0.0.1，手工解析 HTTP 请求行（足够覆盖 OAuth 回调）。
pub struct OauthServer {
    /// 停止句柄（发送信号让监听任务退出）
    stop_handle: OauthServerStopHandle,
    /// 监听任务句柄
    listen_task: Option<JoinHandle<()>>,
    /// 回调结果接收端
    result_rx: Option<oneshot::Receiver<OauthCallbackResult>>,
}

/// 可跨任务触发回调监听停止的句柄。
#[derive(Clone)]
pub struct OauthServerStopHandle {
    stop_tx: watch::Sender<bool>,
}

impl OauthServerStopHandle {
    /// 通知 OAuth server 停止等待回调。
    pub fn stop(&self) {
        let _ = self.stop_tx.send(true);
    }
}

impl OauthServer {
    /// 启动监听。重复调用会返回错误。
    pub async fn start(port: u16) -> AppResult<Self> {
        let addr = format!("{}:{}", constants::LOOPBACK_HOST, port);
        tracing::info!(addr = %addr, "启动 OAuth 回调监听");
        // 仅绑定 loopback IPv4
        let listener = TcpListener::bind(&addr)
            .await
            .map_err(|e| AppError::generic(format!("绑定回调端口失败：{e}")))?;

        let (result_tx, result_rx) = oneshot::channel::<OauthCallbackResult>();
        let (stop_tx, mut stop_rx) = watch::channel(false);
        let stop_handle = OauthServerStopHandle { stop_tx };

        let listen_task = tokio::spawn(async move {
            loop {
                tokio::select! {
                    biased;
                    changed = stop_rx.changed() => {
                        if changed.is_ok() && *stop_rx.borrow() {
                            break;
                        }
                    }
                    accept = listener.accept() => {
                        match accept {
                            Ok((mut stream, _)) => {
                                let result = handle_request(&mut stream).await;
                                // 回写响应页
                                let html = build_response_page(&result);
                                let _ = write_response(&mut stream, &html).await;
                                // 完成回调
                                let _ = result_tx.send(result.clone());
                                // 单次使用：拿到结果后停止监听
                                break;
                            }
                            Err(e) => {
                                tracing::warn!(error = %e, "OAuth 回调 accept 失败");
                            }
                        }
                    }
                }
            }
            tracing::info!("OAuth 回调 server 已关闭");
        });

        Ok(Self {
            stop_handle,
            listen_task: Some(listen_task),
            result_rx: Some(result_rx),
        })
    }

    /// 获取可克隆停止句柄，供取消授权从外部关闭监听。
    pub fn stop_handle(&self) -> OauthServerStopHandle {
        self.stop_handle.clone()
    }

    /// 等待授权码（带超时）。超时返回 [`AppError::auth_timeout`]。
    /// 对齐 dart `OauthServer.waitForCallback`。
    pub async fn wait_for_callback(mut self) -> AppResult<OauthCallbackResult> {
        let result_rx = self
            .result_rx
            .take()
            .ok_or_else(|| AppError::generic("OauthServer 未启动"))?;
        match timeout(
            Duration::from_secs(constants::OAUTH_TIMEOUT_SECS),
            result_rx,
        )
        .await
        {
            Ok(Ok(result)) => {
                self.stop().await;
                Ok(result)
            }
            Ok(Err(_)) => {
                self.stop().await;
                Err(AppError::generic("OAuth 回调通道关闭"))
            }
            Err(_) => {
                self.stop().await;
                tracing::warn!("OAuth 回调等待超时");
                Err(AppError::auth_timeout())
            }
        }
    }

    /// 关闭 server，释放端口。
    pub async fn stop(mut self) {
        self.stop_handle.stop();
        if let Some(handle) = self.listen_task.take() {
            let _ = handle.await;
        }
    }
}

/// 解析 HTTP 请求，提取回调参数。
async fn handle_request(stream: &mut (impl tokio::io::AsyncRead + Unpin)) -> OauthCallbackResult {
    let mut buf = [0u8; 4096];
    let n = stream.read(&mut buf).await.unwrap_or(0);
    let request = String::from_utf8_lossy(&buf[..n]);

    // 解析请求行：GET /oauth/callback?code=xxx&state=yyy HTTP/1.1
    let request_line = request.lines().next().unwrap_or("");
    let path = request_line.split_whitespace().nth(1).unwrap_or("");

    if !path.starts_with(constants::CALLBACK_PATH) {
        return OauthCallbackResult {
            error: Some("无效回调路径".to_string()),
            ..Default::default()
        };
    }

    // 提取 query string
    let query = path.split('?').nth(1).unwrap_or("");
    parse_query(query)
}

/// 解析 query string 为回调结果。
fn parse_query(query: &str) -> OauthCallbackResult {
    let mut result = OauthCallbackResult::default();
    for pair in query.split('&') {
        let mut kv = pair.splitn(2, '=');
        let key = url_decode(kv.next().unwrap_or(""));
        let value = url_decode(kv.next().unwrap_or(""));
        match key.as_str() {
            "code" => result.code = Some(value),
            "state" => result.state = Some(value),
            "error" => result.error = Some(value),
            "error_description" => result.error_description = Some(value),
            "sub_error" => result.sub_error = Some(value),
            _ => {}
        }
    }
    result
}

/// URL 解码（百分号编码 + '+' 当空格，对齐 form-urlencoded / dart Uri.splitQueryString）。
fn url_decode(s: &str) -> String {
    // form-urlencoded 中 '+' 表示空格（percent-decoding 不处理 '+'，需先替换）
    let with_spaces = s.replace('+', " ");
    percent_encoding::percent_decode_str(&with_spaces)
        .decode_utf8_lossy()
        .to_string()
}

/// 写入带长度和关闭连接头的 UTF-8 HTML 成功响应。
async fn write_response(
    stream: &mut (impl tokio::io::AsyncWrite + Unpin),
    html: &str,
) -> std::io::Result<()> {
    let response = format!(
        "HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}",
        html.len(),
        html
    );
    stream.write_all(response.as_bytes()).await?;
    stream.flush().await?;
    Ok(())
}

/// 构建回写浏览器的友好页面。对齐 dart `_buildResponsePage`。
fn build_response_page(result: &OauthCallbackResult) -> String {
    if result.is_success() {
        r#"<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>授权成功</title>
<style>body{font-family:-apple-system,sans-serif;text-align:center;margin-top:80px;color:#333}
h1{color:#1a7f37}</style></head>
<body><h1>✅ 授权成功</h1>
<p>已成功登录华为云盘，现在可以关闭此页面并返回 App。</p></body></html>"#
            .to_string()
    } else {
        let reason = result.error.as_deref().unwrap_or("未知错误");
        format!(
            r#"<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>授权失败</title>
<style>body{{font-family:-apple-system,sans-serif;text-align:center;margin-top:80px;color:#333}}
h1{{color:#d73a49}}</style></head>
<body><h1>❌ 授权失败</h1>
<p>{reason}</p>
<p>请返回 App 重新登录。</p></body></html>"#
        )
    }
}

File: lib/service/auth/oauth_server.dart
/// OAuth 本地回调 HTTP Server（需求 F-AUTH-02 / F-AUTH-06）。
///
/// 严格对齐 Rust 原版 `src/auth/oauth_server.rs`：
/// - 绑定 127.0.0.1:port（不监听 0.0.0.0，满足安全要求）
/// - 监听 GET /oauth/callback，解析 code/state/error/sub_error
/// - 单次使用：拿到首个请求结果后自动关闭
library;

import 'dart:async';
import 'dart:io';

import 'package:petal_link/core/error/app_error.dart';
import 'package:petal_link/core/logger/logger.dart';
import 'package:petal_link/service/auth/auth_constants.dart';

/// OAuth 回调结果（对齐 Rust `OauthCallbackResult`）。
class OauthCallbackResult {
  /// 授权码
  final String? code;

  /// 防 CSRF state
  final String? state;

  /// 华为 error 码（如 '1101'）
  final String? error;

  /// 华为 error_description（如 'invalid scope'）
  final String? errorDescription;

  /// 华为 sub_error（如 '20042' 表示 scope 未授权）
  final String? subError;

  const OauthCallbackResult({
    this.code,
    this.state,
    this.error,
    this.errorDescription,
    this.subError,
  });

  /// 是否成功（有 code 且无 error）
  bool get isSuccess => code != null && error == null;
}

/// 可跨调用方触发回调监听停止的句柄（对齐 Rust `OauthServerStopHandle`）。
class OauthServerStopHandle {
  final void Function() _stop;

  OauthServerStopHandle._(this._stop);

  /// 通知 OAuth server 停止等待回调。
  void stop() => _stop();
}

/// 本地 OAuth 回调服务器。
///
/// 使用 dart:io HttpServer 监听 127.0.0.1；首个到达的请求即为回调结果
/// （含无效路径，对齐 Rust「单次使用」语义）。
class OauthServer {
  final HttpServer _server;
  final StreamSubscription<HttpRequest> _subscription;
  final Completer<OauthCallbackResult> _resultCompleter =
      Completer<OauthCallbackResult>();
  late final OauthServerStopHandle _stopHandle;

  /// 是否已停止（幂等保护）
  bool _stopped = false;

  OauthServer._(this._server, this._subscription) {
    _stopHandle = OauthServerStopHandle._(() => unawaited(stop()));
    // 预挂错误处理：server 在无人等待时被 stop（如授权流程提前失败），
    // 避免 completeError 成为 unhandled async error
    unawaited(_resultCompleter.future.catchError((_) => _closedResult));
  }

  /// 预挂错误处理的占位结果（仅用于让派生 Future 正常完成）
  static const OauthCallbackResult _closedResult =
      OauthCallbackResult(error: '通道已关闭');

  /// 启动监听（绑定失败抛 [AppError]）。重复启动新实例即可，无全局状态。
  static Future<OauthServer> start(int port) async {
    final HttpServer server;
    try {
      AppLogger.i('启动 OAuth 回调监听：${AuthConstants.loopbackHost}:$port');
      // 仅绑定 loopback IPv4
      server = await HttpServer.bind(InternetAddress.loopbackIPv4, port);
    } catch (e) {
      throw AppError.generic('绑定回调端口失败：$e');
    }

    late OauthServer self;
    final subscription = server.listen(
      (request) => self._handleRequest(request),
      onError: (Object e) => AppLogger.w('OAuth 回调 accept 失败：$e'),
    );
    self = OauthServer._(server, subscription);
    return self;
  }

  /// 实际监听端口（port=0 时由系统分配，供测试使用）。
  int get port => _server.port;

  /// 获取停止句柄，供取消授权从外部关闭监听。
  OauthServerStopHandle get stopHandle => _stopHandle;

  /// 等待授权码（带超时）。
  ///
  /// - 正常：返回回调结果并关闭 server
  /// - 超时：抛 `AppError.authTimeout()`
  /// - 外部 stop（用户取消）：抛 `AppError.generic('OAuth 回调通道关闭')`
  ///
  /// 对齐 Rust `OauthServer.wait_for_callback`。
  Future<OauthCallbackResult> waitForCallback({
    Duration timeout = AuthConstants.oauthTimeout,
  }) async {
    try {
      final result = await _resultCompleter.future.timeout(timeout);
      await stop();
      return result;
    } on TimeoutException {
      await stop();
      AppLogger.w('OAuth 回调等待超时');
      throw AppError.authTimeout();
    } catch (_) {
      await stop();
      rethrow;
    }
  }

  /// 关闭 server，释放端口（幂等）。
  Future<void> stop() async {
    if (_stopped) return;
    _stopped = true;
    // 先释放等待方（取消语义即时生效）：通道关闭语义（对齐 Rust oneshot 关闭）
    if (!_resultCompleter.isCompleted) {
      _resultCompleter.completeError(AppError.generic('OAuth 回调通道关闭'));
    }
    await _subscription.cancel();
    try {
      // 非强制关闭：等在途响应发完；超时兜底强制关闭（防御浏览器预连接挂起）
      await _server.close().timeout(
        const Duration(seconds: 2),
        onTimeout: () => _server.close(force: true),
      );
    } catch (_) {
      // 关闭异常不影响结果
    }
    AppLogger.i('OAuth 回调 server 已关闭');
  }

  /// 处理首个 HTTP 请求：解析回调参数 → 回写响应页 → 完成结果并关闭。
  Future<void> _handleRequest(HttpRequest request) async {
    final result = _parseRequest(request);
    try {
      request.response
        ..statusCode = HttpStatus.ok
        ..headers.contentType =
            ContentType('text', 'html', charset: 'utf-8')
        ..headers.set(HttpHeaders.connectionHeader, 'close')
        ..write(buildResponsePage(result));
      await request.response.close();
    } catch (e) {
      AppLogger.w('OAuth 回写响应页失败', e);
    }
    if (!_resultCompleter.isCompleted) {
      _resultCompleter.complete(result);
    }
    // 单次使用：拿到结果后停止监听
    await stop();
  }

  /// 解析请求，提取回调参数（对齐 Rust `handle_request` + `parse_query`）。
  ///
  /// Dart `Uri.queryParameters` 按 form-urlencoded 解码（'+' 视为空格），
  /// 与 Rust `url_decode` 语义一致。
  OauthCallbackResult _parseRequest(HttpRequest request) {
    if (request.uri.path != AuthConstants.callbackPath) {
      return const OauthCallbackResult(error: '无效回调路径');
    }
    final q = request.uri.queryParameters;
    String? pick(String key) {
      final v = q[key];
      return v == null || v.isEmpty ? null : v;
    }

    return OauthCallbackResult(
      code: pick('code'),
      state: pick('state'),
      error: pick('error'),
      errorDescription: pick('error_description'),
      subError: pick('sub_error'),
    );
  }
}

/// 构建回写浏览器的友好页面（对齐 Rust `build_response_page`）。
String buildResponsePage(OauthCallbackResult result) {
  if (result.isSuccess) {
    return '<!DOCTYPE html>\n'
        '<html><head><meta charset="utf-8"><title>授权成功</title>\n'
        '<style>body{font-family:-apple-system,sans-serif;text-align:center;margin-top:80px;color:#333}\n'
        'h1{color:#1a7f37}</style></head>\n'
        '<body><h1>✅ 授权成功</h1>\n'
        '<p>已成功登录华为云盘，现在可以关闭此页面并返回 App。</p></body></html>';
  }
  final reason = result.error ?? '未知错误';
  return '<!DOCTYPE html>\n'
      '<html><head><meta charset="utf-8"><title>授权失败</title>\n'
      '<style>body{font-family:-apple-system,sans-serif;text-align:center;margin-top:80px;color:#333}\n'
      'h1{color:#d73a49}</style></head>\n'
      '<body><h1>❌ 授权失败</h1>\n'
      '<p>$reason</p>\n'
      '<p>请返回 App 重新登录。</p></body></html>';
}
